bothan-binance/src/worker/asset_worker.rs

```diff
@@ -0,0 +1,165 @@
+use std::sync::Weak;
+
+use rust_decimal::Decimal;
+use tokio::select;
+use tokio::sync::mpsc::Receiver;
+use tokio::time::{sleep, timeout};
+use tracing::{debug, error, info};
+
+use bothan_core::store::Store;
+use bothan_core::types::AssetInfo;
+
+use crate::api::msgs::{BinanceResponse, Data};
+use crate::api::{self, BinanceWebSocketConnection, BinanceWebSocketConnector};
+use crate::worker::error::ParseError;
+use crate::worker::types::{DEFAULT_TIMEOUT, RECONNECT_BUFFER};
+use crate::worker::BinanceWorker;
+
+pub(crate) async fn start_asset_worker(
+    worker: Weak<BinanceWorker>,
+    mut connection: BinanceWebSocketConnection,
+    mut subscribe_rx: Receiver<Vec<String>>,
+    mut unsubscribe_rx: Receiver<Vec<String>>,
+) {
+    loop {
+        select! {
+            Some(id) = subscribe_rx.recv() => handle_subscribe_recv(id, &mut connection).await,
+            Some(id) = unsubscribe_rx.recv() => handle_unsubscribe_recv(id, &mut connection).await,
+            result = timeout(DEFAULT_TIMEOUT, connection.next()) => {
+                if let Some(worker) = worker.upgrade() {
+                    match result {
+                        Err(_) => handle_reconnect(&worker.connector, &mut connection, &worker.store).await,
+                        Ok(binance_result) => handle_connection_recv(binance_result, &worker.connector, &mut connection, &worker.store).await,
+                    }
+                } else {
+                    break
+                }
+            }
+        }
+    }
+    debug!("asset worker exited")
+}
+
+async fn subscribe(
+    ids: Vec<String>,
+    connection: &mut BinanceWebSocketConnection,
+) -> Result<(), api::SubscriptionError> {
+    if !ids.is_empty() {
+        connection
+            .subscribe_mini_ticker_stream(&ids.iter().map(|s| s.as_str()).collect::<Vec<&str>>())
+            .await?
+    }
+
+    Ok(())
+}
+
+async fn handle_subscribe_recv(ids: Vec<String>, connection: &mut BinanceWebSocketConnection) {
+    if let Err(e) = subscribe(ids, connection).await {
+        error!("failed to subscribe: {}", e);
+    } else {
+        info!("subscribed to ids");
+    }
+}
+
+async fn handle_unsubscribe_recv(ids: Vec<String>, connection: &mut BinanceWebSocketConnection) {
+    if ids.is_empty() {
```

Comment:
Why we specially handle empty vec unsub here and not above?

Reply:
its handled in subscribe() but will refactor to increase readability